### PR TITLE
Avoid ToUpper for string comparison

### LINKED ISSUE #8629: Avoid ToUpper for string comparison
`ToUpper()` should not be used for string comparison but we do so in our code base.
 
> The [ToUpper](https://learn.microsoft.com/en-us/dotnet/api/system.string.toupper?view=net-8.0) method is often used to convert a string to uppercase so that it can be used in a case-insensitive comparison. A better method to perform case-insensitive comparison is to call a string comparison method that has a [StringComparison](https://learn.microsoft.com/en-us/dotnet/api/system.stringcomparison?view=net-8.0) parameter whose value you set to [StringComparison.CurrentCultureIgnoreCase](https://learn.microsoft.com/en-us/dotnet/api/system.stringcomparison?view=net-8.0#system-stringcomparison-currentcultureignorecase) for a culture-sensitive, case-insensitive comparison.

[Source](https://learn.microsoft.com/en-us/dotnet/api/system.string.toupper?view=net-8.0#system-string-toupper(system-globalization-cultureinfo))

See also [Best practices for comparing strings in .NET](https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#stringtoupper-and-stringtolower)

We should replace the following code snippets.

https://github.com/SonarSource/sonar-dotnet/blob/3ca63c72861e4de6a7fe110384258bbe61e47487/analyzers/src/SonarAnalyzer.Common/Common/AnalyzerLanguage.cs#L54

https://github.com/SonarSource/sonar-dotnet/blob/3ca63c72861e4de6a7fe110384258bbe61e47487/analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs#L97

https://github.com/SonarSource/sonar-dotnet/blob/3ca63c72861e4de6a7fe110384258bbe61e47487/analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs#L116

https://github.com/SonarSource/sonar-dotnet/blob/3ca63c72861e4de6a7fe110384258bbe61e47487/analyzers/src/SonarAnalyzer.Common/Rules/FieldShadowsParentFieldBase.cs#L51-L68

https://github.com/SonarSource/sonar-dotnet/blob/3ca63c72861e4de6a7fe110384258bbe61e47487/analyzers/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs#L51

see also https://github.com/SonarSource/sonar-dotnet/pull/8586

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Fixes #8629

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs [hunk 1]
```diff
@@ -24,32 +24,26 @@ namespace SonarAnalyzer.Helpers
...
+            "<autogenerated",
+            "generated by");
+
+        private static readonly ImmutableList<string> GeneratedCodeAttributes = ImmutableList.Create(
+            "DebuggerNonUserCode",
+            "DebuggerNonUserCodeAttribute",
+            "GeneratedCode",
+            "GeneratedCodeAttribute",
+            "CompilerGenerated",
+            "CompilerGeneratedAttribute");
```
Comment (Tim-Pohlmann) [inline analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs:46]: Why the change to `ImmutableList`? `string[]` seems sufficient in this case.

File: analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs [hunk 2]
```diff
@@ -92,11 +86,7 @@ private bool HasAutoGeneratedComment(SyntaxNode root)
 
             return firstToken.LeadingTrivia
                 .Where(IsTriviaComment)
-                .Any(trivia =>
-                {
-                    var commentText = trivia.ToString().ToUpperInvariant();
-                    return Array.Exists(AutoGeneratedCommentParts, commentText.Contains);
-                });
+                .Any(trivia => AutoGeneratedCommentParts.Exists(part => trivia.ToString().IndexOf(part, StringComparison.OrdinalIgnoreCase) >= 0));
```
Comment (Tim-Pohlmann) [inline analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs:89]: Use [Contains](https://learn.microsoft.com/en-us/dotnet/api/system.string.contains?view=net-8.0#system-string-contains(system-string-system-stringcomparison)) instead of `IndexOf`.

File: analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs (hunk 1 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs:46]: And if you really want it to be immutable, why not `ImmutableArray`?

**Review by Tim-Pohlmann — CHANGES_REQUESTED** [review]
Some questions

File: analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs (hunk 2 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs:89]: That is not possible. That overload is `netstandard2.1`. In the Remarks section, `IndexOf` is the recommended alternative.

**Review by Tim-Pohlmann — APPROVED** [review]
LGTM except for the code coverage.
